packages/gatsby-transformer-csv/src/gatsby-node.js

```diff
@@ -25,33 +32,62 @@ async function onCreateNode(
   }
   // Load CSV contents
   const content = await loadNodeContent(node)
+
+  // Destructure out our custom options
+  const { typeName, nodePerFile, ...options } = pluginOptions || {}
+
   // Parse
   let parsedContent = await convertToJson(content, options)
 
+  // Generate the type
+  function getType({ node, object }) {
+    if (pluginOptions && _.isFunction(typeName)) {
+      return pluginOptions.typeName({ node, object })
+    } else if (pluginOptions && _.isString(typeName)) {
+      return _.upperFirst(_.camelCase(typeName))
+    } else {
+      return typeNameFromFile({ node })
+    }
+  }
+
+  // Generate the new node
+  function transformObject(obj, i) {
+    const csvNode = {
+      ...obj,
+      id: obj.id ? obj.id : createNodeId(`${node.id} [${i}] >>> CSV`),
+      children: [],
+      parent: node.id,
+      internal: {
+        contentDigest: createContentDigest(obj),
+        // TODO make choosing the "type" a lot smarter. This assumes
+        // the parent node is a file.
+        // PascalCase
+        type: getType({ node, object: parsedContent }),
+      },
+    }
+
+    createNode(csvNode)
+    createParentChildLink({ parent: node, child: csvNode })
+  }
+
   if (_.isArray(parsedContent)) {
-    const csvArray = parsedContent.map((obj, i) => {
-      return {
-        ...obj,
-        id: obj.id ? obj.id : createNodeId(`${node.id} [${i}] >>> CSV`),
-        children: [],
-        parent: node.id,
-        internal: {
-          contentDigest: createContentDigest(obj),
-          // TODO make choosing the "type" a lot smarter. This assumes
-          // the parent node is a file.
-          // PascalCase
-          type: _.upperFirst(_.camelCase(`${node.name} Csv`)),
-        },
+    if (pluginOptions && nodePerFile) {
+      if (pluginOptions && _.isString(nodePerFile)) {
+        transformObject({ [nodePerFile]: parsedContent }, 0)
+      } else {
+        transformObject({ items: parsedContent }, 0)
       }
-    })
-
-    _.each(csvArray, y => {
-      createNode(y)
-      createParentChildLink({ parent: node, child: y })
-    })
+    } else {
+      _.each(parsedContent, (obj, i) => {
+        transformObject(obj, i)
+      })
+    }
   }
 
   return
 }
 
+exports.typeNameFromDir = typeNameFromDir
+exports.typeNameFromFile = typeNameFromFile
```

Comment:
I think we need to move those functions and exports to `index.js` file:
 - this is default entrypoint for packages (otherwise ``const { typeNameFromDir } = require(`gatsby-transformer-csv`)`` probably won't work
 - gatsby will warn (or error out) when gatsby-node export names doesn't match gatsby-node API extension points